Global style revisions: show change summary on selected item

ramonjd opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

Maybe one day resolves https://github.com/WordPress/gutenberg/issues/55647

Part of:

- https://github.com/WordPress/gutenberg/issues/55776



https://github.com/WordPress/gutenberg/assets/6458278/0c57e6f5-0cf7-40f1-a750-ec40dbbbfd92




## Why?

To give the user a short overview of the global styles revision change set, which may not be evident from the canvas preview.

Also, highlighting revisions that match current editor styles by a left-hand-side dark, border.

## TODO

- [x] Unit tests

## How?

1. Deep comparison of two objects to get changed values
2. Limiting reported changes to the second level, e.g., `Block name > style`
3. Deduping results and translating
4. Caching results 

## Testing Instructions

Make a bunch of changes to your global styles revisions, ensuring to get a decent mix of block, top-level, element styles and layout, typography and color settings. 

If the theme, e.g., 2024, has style variations, activate and save a few revision using those.

Open up the styles revisions panel and check that the change summary makes sense.

ramonjd committed:
Moving date format setting call into the comoponent
rejigging getLabel function
adding css var

ramonjd committed:
Testing a message to indicate that the revisions state is the same as the editor state.

ramonjd committed:
Working on change list, adding translations. WIP

ramonjd committed:
Adding more translations.

ramonjd committed:
Revert button-in-item for another day
Reduce depth of changeset and remove unused translations
Display aria-label on button instead of tooltip

andrewserong commented on packages/edit-site/src/components/global-styles/screen-revisions/get-revision-changes.js:
```diff
@@ -0,0 +1,174 @@
+	'color.background': __( 'background colors' ),
+	'spacing.margin': __( 'margin styles' ),
+	'spacing.padding': __( 'padding styles' ),
+	'spacing.blockGap': __( 'block gap' ),
```
Tiny nit: should this be `block spacing` as the text string, to match the UI control?

ramonjd replied:
Good call!

andrewserong commented on packages/edit-site/src/components/global-styles/screen-revisions/get-revision-changes.js:
```diff
@@ -0,0 +1,174 @@
+	'spacing.margin': __( 'margin styles' ),
+	'spacing.padding': __( 'padding styles' ),
+	'spacing.blockGap': __( 'block gap' ),
+	'settings.layout': __( 'layout settings' ),
```
From a user perspective, the only things that will have changed here is content or wide size, I think? So `settings` might not feel very direct to the user. We could either split this out into content/wide size, or use `layout styles` for consistency?

Another idea, though, could be to remove the word `style` or `styles` from each of these strings, and use `style` or plural `styles` at the end of the text string?

ramonjd replied:
Oh yes, fair point. I like `layout styles`.

> could be to remove the word style or styles from each of these strings, and use style or plural styles at the end of the text string?

I like this idea. I'll have a play around with it - I just need to make sure it plays well with translatable strings. If not, I think having "styles" after each higher-level style group is okay.

andrewserong commented on packages/edit-site/src/components/global-styles/screen-revisions/get-revision-changes.js:
```diff
@@ -0,0 +1,174 @@
+	);
+
+	// Remove dupes and shuffle results.
+	const result = shuffle( [ ...new Set( changedValueTree ) ] )
```
Why do results get shuffled? Apologies, I think this might have been mentioned elsewhere!

ramonjd replied:
Very good question. 

It's a bit of an artefact now that we've reduced the depth of reported changes. I left it in there in case I needed it again. 

I mainly put it there to ensure we saw a mix of changes from a large or repetitive change set. The comparison function starts at the top and works to the bottom. 

Before, even at three levels deep, the change set could have contained `n` x font size settings, or `n` x color palette changes. Under styles, the same styles could have changed over `n` revisions.

So shuffling the order prevented us rendering either just settings changes (because they're always at the start of the array due to their location in the global styles collection) or the same thing over and over again.

I think we can get rid of it, and make a call as to whether we need to reinstate to spice up the results.

andrewserong reviewed (commented):
This feel pretty promising to me! It's quite nice even having a few words when clicking between each revision.

One thought I had was whether it'd be possible to drop `style` or `styles` or `color` and `colors` and group by type, if it helps shorten things a little? Not a worry if that would make things too complex, though.

But it's already feeling pretty good to me so far 🙂

<img width="267" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/408a9993-8054-492c-9287-8072b7f32b7f">

ramonjd committed:
Removing shuffle function and fixing up block spacing translation

ramonjd commented on packages/edit-site/src/components/global-styles/screen-revisions/get-revision-changes.js:
```diff
@@ -0,0 +1,166 @@
+	maxResults = 5
+) {
+	if ( cache.has( revision.id ) ) {
+		return cache.get( revision.id );
```
Note to self: look into how to cache "unsaved" changes, which don't have a unique id. Maybe hash the stringified JSON or something.

ramonjd replied:
Done this by using the revision has the Map key. This allows us to cache the revision sets themselves, and therefore account for changes to Unsaved revisions.

Not 100% sure on performance yet.

ramonjd committed:
Using the revision has the Map key. This allows us to cache the revision sets themselves and account for changes to Unsaved revisions.

ramonjd committed:
Used WeakMap in favour of Map for garbage collection, if it helps at all

ramonjd committed:
Remove hasMore var - unneeded because it's only used once

ramonjd committed:
Tidying up - remove .map loop

ramonjd committed:
getGlobalStylesChanges was doing nothing! Removed.

ramonjd committed:
Using revision + previousRevision combo for cache key to ensure that the results are cached for the same two objects
Returning from cache where maxResults value is smaller than cached results
Added first tests

ramonjd committed:
Moving maxResults decisions to consuming component. getRevisionChanges returns an unadulterated array.

ramonjd commented on packages/edit-site/src/components/global-styles/screen-revisions/get-revision-changes.js:
```diff
@@ -0,0 +1,161 @@
+	if ( ! isObject( changedObject ) && ! isObject( originalObject ) ) {
+		// Only return a path if the value has changed.
+		// And then only the path name up to 2 levels deep.
+		return changedObject !== originalObject
```
So this is very specific to the UI's requirements.

We're just returning the path to a value that has changed between versions.

If we want to reuse this function later, we can return anything, e.g., an object:

```js
{
    path: 'styles.typography.fontWeight',
    oldValue: 600,
    newValue: 300,
}
```

ramonjd committed:
Move get blockNames to main component

ramonjd committed:
Have to use map because WeakMap wants the same reference as the object key.

ramonjd commented on packages/edit-site/src/components/global-styles/screen-revisions/get-revision-changes.js:
```diff
@@ -0,0 +1,161 @@
+	previousRevision,
+	blockNames
+) {
+	const cacheKey = JSON.stringify( { revision, previousRevision } );
```
This might be an overkill, but it's the only way I think of to ensure that we can return the right cached results for a matching set of `revision` and `previousRevision`.

ramonjd committed:
Remove the trailing comma on truncated results

apeatling reviewed (commented):
This is looking good. I think it's a nice improvement and gives that extra bit of information.

I do wonder if we need to have some sort of context as to what the words mean. Do you think saying "Styles changed: xyz" would be better?

I noticed a couple of things. The text should line up and not be inset slightly. Also does the ellipsis mean data has been truncated?

<img width="375" alt="Screenshot 2023-12-06 at 12 39 33 PM" src="https://github.com/WordPress/gutenberg/assets/1464705/99d58fb3-ccdd-4657-92a1-eaf104c759b4">

apeatling commented:
I see it does mean it's truncated `if ( changesLength > MAX_CHANGES ) {`

ramonjd commented:
@apeatling What do you think about this?

<img width="258" alt="Screenshot 2023-12-07 at 12 25 48 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/4be0f0cb-11b2-4487-a64e-d3689dbdbc38">

Forgetting everything else in that screenshot, maybe it'd be useful to somewhere show the truncated results and how many more there are (`and n more...`)?

andrewserong commented:
Oh, I like that list idea, makes it easier to see changes at a glance compared to a paragraph / sentence based structure.

andrewserong commented on packages/edit-site/src/components/global-styles/screen-revisions/get-revision-changes.js:
```diff
@@ -0,0 +1,161 @@
+
+/**
+ * Get an array of translated summarized global styles changes.
+ * Results are cached using a WeakMap key of `{ revision, previousRevision }`.
```
Tiny nit: this mentions `WeakMap` but at the beginning of the file, it appears to be a `Map`.

ramonjd replied:
Good spotting. 👍🏻  I swapped from WeakMap to Map because of the change in key.

andrewserong commented:
Just another idea, if we wanted to find a mid-point between a list and sentence structure (if a list winds up being too long), could we kind of group them by type, to remove repetition of the word `Block` or `Styles`? Here's a quick browser mockup, not sure if it's any good as an idea: 🤔 

<img width="259" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/3fa78dd9-0e16-4f63-b2a9-79244d0849c5">

ramonjd commented on packages/edit-site/src/components/global-styles/screen-revisions/get-revision-changes.js:
```diff
@@ -0,0 +1,161 @@
+
+/**
+ * Get an array of translated summarized global styles changes.
+ * Results are cached using a WeakMap key of `{ revision, previousRevision }`.
```
```suggestion
 * Results are cached using a Map() key of `JSON.stringify( { revision, previousRevision } )`.
```

ramonjd committed:
Test commit: listing changes, showing `and n more`

ramonjd committed:
Test commit: grouping changes using tuples

ramonjd commented:
I have a couple of commits lined up. Here's the one with the grouping, which required a bit of a code refactor, but simplifies the translations a lot.

https://github.com/WordPress/gutenberg/assets/6458278/6ea8f5c3-d394-455f-801f-5ae8e9b13ae9

And without a heading:

<img width="253" alt="Screenshot 2023-12-07 at 3 14 05 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/8f7a4f3d-f027-4c30-bc0c-15edf5fa7cf3">

It's all tweakable 😄 

And the other commit [mentioned above](https://github.com/WordPress/gutenberg/pull/56577#issuecomment-1843984443).

ramonjd commented on packages/edit-site/src/components/global-styles/screen-revisions/revisions-buttons.js:
```diff
@@ -6,28 +6,91 @@ import classnames from 'classnames';
+	}, {} );
+
+	return (
+		<div className="edit-site-global-styles-screen-revision__changes">
```
Note to self: probably can do with a little clean up, maybe rethink HTML elements.

andrewserong commented:
> And without a heading:

Nice, that's probably my favourite so far! Sounds like this PR might be ready for design feedback now?

ramonjd commented:
> Nice, that's probably my favourite so far! Sounds like this PR might be ready for design feedback now?

Thanks for testing! Yeah, I'm now in UI bike shedding territory 😄

jameskoster commented:
Thanks for the ping, and nice work so far.

My initial thought is that we probably don't need to be so granular for typography, styles, and colors. It might be enough to simply list each one accordingly which would reduce a lot of noise. In a similar vein I'd agree with removing the heading. Blocks can continue to be listed. 

> Saved December 7, 2023
> Typography, Colors, Layout, Quote block, Paragraph block, Image block, Cover block, Buttons block, Button block
> Admin

What do you think?

apeatling commented:
Thanks for the effort here @ramonjd. I gave the current version a spin, and had a look at your efforts above. To be honest I think with the extra detail the scan-ability is lost, and adds extra noise like @jameskoster mentioned.

I think what we had originally is actually better, and we can see if the context issue comes up as an actual problem once it's out there.

ramonjd commented:
> It might be enough to simply list each one accordingly which would reduce a lot of noise. In a similar vein I'd agree with removing the heading. Blocks can continue to be listed.
> I think what we had originally is actually better, and we can see if the context issue comes up as an real problem once it's out there.

This helps a lot, thanks, folks.

Easy to revert since it's what we started with, and I'd suspect it'd be a little more performant on the client side too with not complicated component rendering.

Cheers!

annezazu commented:
+1 on reducing noise for now and iterating as feedback comes in!

ramonjd committed:
Reverting back to comma-separate list of changes
Added e2e assertion

ramonjd commented on packages/edit-site/src/components/global-styles/screen-revisions/index.js:
```diff
@@ -215,14 +216,6 @@ function ScreenRevisions() {
 					) }
 				</>
-			) : (
-				<Spacer marginX={ 4 } data-testid="global-styles-no-revisions">
```
This was dead code because we never show the panel where there are no revision.

ramonjd commented:
I think I've addressed all the latest feedback, thanks for your input 🙇🏻

ramonjd commented:
Also wondering, would it be useful to have a link to `revison.php` on each item? It's cheap to do, and maybe useful to some:

<img width="662" alt="Screenshot 2023-12-08 at 11 17 52 am" src="https://github.com/WordPress/gutenberg/assets/6458278/810c5284-ae5f-4139-b159-4012c642efad">

ramonjd committed:
Swapping order of author name and changes block
Moving everything into the button so it's clickable.

ramonjd commented:
> The author can appear beneath the change detail.

Done that and updated PR description 🙇🏻

ramonjd committed:
Don't live in the past, man

jameskoster commented:
> So something like Paragraph block: typography, color...? I think that's very doable.

No, not quite. I'm suggesting to align the listed changes with the sections titles in the Styles panel. So if there are any changes in the "Typography" panel then the revision simply lists "Typography" as a change. Same for Colors and Layout. For blocks, each edited block is listed. 

So if I changed the Link element, adjusted global block spacing, and added a border radius to the Image block the change list would be: "Typography, Layout, Image block".

> Maybe we can leave it it out and pursue this in a follow up.

I think that's a good idea, it needs a little design exploration.

ramonjd commented:
> I'm suggesting to align the listed changes with the sections titles in the Styles panel. 

I get you, thanks for the clarification. I think this also should be doable 😄

apeatling reviewed (approved):
Gave this another run through. I think this is looking good for an initial effort and is a definite improvement, so let's get it in and we can iterate more from here. :+1:

<img width="322" alt="Screenshot 2023-12-12 at 1 11 36 PM" src="https://github.com/WordPress/gutenberg/assets/1464705/539985e8-f2c1-42d5-b08f-09e0fc1d7824">

andrewserong reviewed (approved):
> I think this is looking good for an initial effort and is a definite improvement, so let's get it in and we can iterate more from here. 👍

+1 this looks like a great place to merge for now and see how it feels for folks. Definitely feels nice to me in its current form, and the code looks good!

https://github.com/WordPress/gutenberg/assets/14988353/37d98e05-0199-4c7c-8ded-c3ae9d2516c6

✨

ramonjd commented:
Thanks again for sticking with me on this one, folks 🙇🏻
